ui/apps/pmm-compat/src/theme.ts

```diff
@@ -20,19 +18,154 @@ export const changeTheme = async (themeId: 'light' | 'dark'): Promise<void> => {
+
+const parseOrigin = (u: string | URL | null | undefined): string | null => {
+  try {
+    const url = typeof u === 'string' ? new URL(u) : u instanceof URL ? u : null;
+    return url ? `${url.protocol}//${url.host}` : null;
+  } catch {
+    return null;
+  }
+};
+
+/**
+ * Resolve initial target origin (may be '*' in dev).
+ * - Dev: start with '*' to support split hosts/ports (vite + docker).
+ * - Prod: concrete origin (document.referrer → window.location.origin).
+ */
+function resolveInitialTargetOrigin(): string {
+  const loc = new URL(window.location.href);
+  if (isDevHost(loc.hostname)) {
+    return '*';
+  }
+  const ref = parseOrigin(document.referrer);
+  return ref ?? `${loc.protocol}//${loc.host}`;
+}
+
+/** Safely obtain a Window to post to (top if cross-framed, otherwise parent/self). */
+function resolveTargetWindow(): Window | null {
+  try {
+    if (window.top && window.top !== window) {
+      return window.top;
+    }
+    if (window.parent) {
+      return window.parent;
+    }
+  } catch (err) {
+    console.warn('[pmm-compat] Failed to send handshake:', err);
+  }
+  return window;
+}
+
+/** Runtime-locked origin (handshake will tighten '*' in dev). */
+const targetOriginRef = { current: resolveInitialTargetOrigin() };
+let lastSentMode: 'light' | 'dark' | null = null;
+
+/** Send helper that always uses the current locked origin. */
+function sendToParent(msg: unknown) {
+  const w = resolveTargetWindow();
+  if (!w) {
+    return;
+  }
+  w.postMessage(msg, targetOriginRef.current);
+}
+
+/** Dev-only handshake: lock '*' to the real origin after the first ACK. */
+(function setupOriginHandshake() {
+  const isDev = isDevHost(new URL(window.location.href).hostname);
+  if (!isDev || targetOriginRef.current !== '*') {
+    return;
+  }
+
+  // Ask parent for its origin once
+  try {
+    sendToParent({ type: 'pmm.handshake' });
+  } catch {
+    // ignore
+  }
+
+  const onAck = (e: MessageEvent<{ type?: string }>) => {
+    if (e?.data?.type !== 'pmm.handshake.ack') {
+      return;
+    }
+    // Lock to the explicit origin provided by the parent
+    targetOriginRef.current = e.origin || targetOriginRef.current;
+    window.removeEventListener('message', onAck);
+  };
+  window.addEventListener('message', onAck);
+})();
+
+// Initial apply from current Grafana theme and notify parent once
+(function initThemeBridge() {
+  const initial: 'light' | 'dark' = config?.theme2?.colors?.mode === 'dark' ? 'dark' : 'light';
+  const mode = applyHtmlTheme(initial);
+  try {
+    if (lastSentMode !== mode) {
+      sendToParent({ type: 'grafana.theme.changed', payload: { mode } });
+      lastSentMode = mode;
+    }
+  } catch (err) {
+    console.warn('[pmm-compat] failed to post initial grafana.theme.changed:', err);
+  }
+})();
+
+// React to Grafana ThemeChangedEvent (Preferences change/changeTheme())
+getAppEvents().subscribe(ThemeChangedEvent, (evt: unknown) => {
```

Comment:
You're subscribing to theme changes on multiple places each with a different handling why is that?